osf/utils/registrations.py

```diff
@@ -0,0 +1,174 @@
+def get_value_or_extra(nested_response, block_type, key, keys):
+    """
+    Sometimes the relevant information is stored under "extra" for files,
+    otherwise, "value".
+
+    :params, nested dictionary
+    :block_type, string, current block type
+    :key, particular key in question
+    :keys, array of keys remaining to recurse through to find the user's answer
+    :returns array (files or multi-response answers) or a string IF deepest level of nesting,
+    otherwise, returns a dictionary to get the next level of nesting.
+    """
+    keyed_value = nested_response.get(key, '')
+    # No guarantee that the key exists in the dictionary
+    if isinstance(keyed_value, basestring):
+        return keyed_value
+
+    # If we are on the most deeply nested key (no more keys left in array),
+    # and the block type is "file-input", the information we want is
+    # stored under extra
+    if block_type == 'file-input' and not keys:
+        extra = format_extra(keyed_value.get('extra', []))
+        return extra
+    else:
+        value = keyed_value.get('value', '')
+        return value
+
+# For flatten_registration_metadata
+def get_nested_answer(nested_response, block_type, keys):
+    """
+    Recursively fetches the nested response in registered_meta.
+
+    :params nested_response dictionary
+    :params keys array, of nested question_ids: ["recommended-analysis", "specify", "question11c"]
+    :returns array (files or multi-response answers) or a string
+    """
+    if isinstance(nested_response, dict):
+        key = keys.pop(0)
+        # Returns the value associated with the given key
+        value = get_value_or_extra(nested_response, block_type, key, keys)
+        return get_nested_answer(value, block_type, keys)
+    else:
+        # Once we've drilled down through the entire dictionary, our nested_response
+        # should be an array or a string
+        return nested_response
+
+# For expanding registration_responses
+def set_nested_values(nested_dictionary, keys, value):
+    """
+    Drills down through the nested dictionary, accessing each key in the array,
+    and sets the last key equal to the  passed in value, if this key doesn't already exist.
+
+    Assumes all keys are present, except for potentially the final key.
+
+    :param nested_dictionary, dictionary
+    :param keys, array
+    :param value, object, array, or string
+    """
+    for key in keys[:-1]:
+        nested_dictionary = nested_dictionary.get(key, None)
+
+    final_key = keys[-1]
+    if not nested_dictionary.get(final_key):
+        nested_dictionary[final_key] = value
+
+# For expanding registration_responses
+def build_answer_block(block_type, value):
+    extra = []
+    if block_type == 'file-input':
+        extra = value
+        value = ''
+        for file in extra:
+            if file.get('file_name'):
+                file['data'] = {
+                    'name': file.get('file_name')  # What legacy FE needs for rendering
```

Comment:
I didn't do enough digging to see what was needed :/ `name` is all that's needed to display the file's name when it's in draft mode, but more is needed to actually register it. Digging into the minimum fields needed on the draft when the node is archived.

Reply:
ok, after some more digging, `file['data']['file_name']` is used for just rendering the file name in the legacy UI on the _draft registration_, but I still needed the pieces for the registration:

Long story short, the `viewUrl` is [built for you](https://github.com/CenterForOpenScience/osf.io/blob/develop/website/archiver/utils.py#L310) on the `Registration` as part of the archiving process, but it needs the `sha256` from the original file to do so.  I'm thinking this is another piece that we should pass in and store on the DraftRegistration.  In addition to `file_name` and `file_id`, we should pass in the `sha256` in `registration_responses`.

Reply:
@aaxelb, can front end pass in `sha256` in addition to the file_name and file_id?  Since you will be making requests to WB, I assume, I think you'll already have that info on the front end?

(This mirrors what is happening on current front legacy front end, which is passing in the sha256 from the WB request).

Reply:
Looked into it with @fabmiz and yes, it should be easy to include `sha256`, let's plan on that. Looking into the future, seems important for a registry to include hashes so anyone can verify the registered data -- and the burden of archiving arbitrary data can more easily be handled by a separate archive.